luizfiorentino/my-city-app
# Location feature

# What this pull request does

## What

In the issue form, the user can choose either to use the current location or to pin the location in the city's map.
In addition, the user can optionally fill in an email address to receive a notification with status and message every time a new update happens.

## Why

- Adding a map with pin on the issue's location is a cool feature, it gives the user an enhanced feeling of interactivity. 
The choice between current and "pin on the map" location ensures the respect for the user's privacy

- Getting an update via email every time a new update occurs gives the user a sense of esteem with regards to the posted issue, showing commitment from the admin to take care of and solve the issue, keeping the user informed about the steps taken by the Municipality

## How

- the map is displayed using `MapContainer`, `TileLayer`, `Marker`, `useMapEvents` from `react-leaflet` library
- user current location is determined with "navigator.geolocation.getCurrentPosition" method 
- to set a location in the city's map, a `Location Marker` function is configured  with the `map.flyTo` method for every click event (user pinning a certain point in the map)
- there is a call to the Open Cage Geocoding api via the Next.js `/api/geolocation` endpoint passing latitude and longitude, and it returns the actual address to be used in the issue card (Admin List) and in the Details Page
- whitelisting and rate limiting rules are added to avoid unnecessary api calls to Open Cage
- there is a migration in the prisma database schedule so that the `Issue` model contains a new field `email` to store the user's provided email
- the user form contains a new input field `email` to store this value (and the correspondent `Zod validation` for the email (optional but if entered must be in the form of an email address)
- when a new update is added by the admin (api call to `/api/statusChanges` endpoint, the user email is passed along with the other infos (status, message, and issueId). The handler first checks whether there is an email address and if there is, it uses `nodemailer` library to send the email using `createTransport` and `sendMail` methods. If there is no email linked to the Issue, it just sends the POST request to insertion on the database


## Steppings to test this

- go to the `/` (form page): submit an issue. You can either do it with or without entering an email address. If you choose to, try something different from an email valid address, and the error message will be displayed
- after filling in a valid email address, and submitting the issue, go to the `/admin` page, click on the just submitted issue's status that leads you to the Details Page
- on the Details Page, click `Edit` select a new status if necessary, and type an update message (the same will should work with the `Mark as solved` button
- open your email account and check that an email from `my-city-app` with the update is sent to inform you about it!

## Anything else

Maybe sending the user an email in case the issue is deleted with a corresponding message?
